### PR TITLE
Add diagonal enum for RectangleMesh

### LINKED ISSUE #1833: Change diagonal argument to enum in mesh construction
We should always avoid the usage of strings. used in `RectangleMesh`
https://github.com/FEniCS/dolfinx/blob/c30adac64e5f5a19e7909666d18a8d6d8d38ebfb/cpp/dolfinx/generation/RectangleMesh.h#L34-L37

### PR DESCRIPTION
**Pull request by jorgensd**
Fixes #1833.

### CODE REVIEW

File: cpp/dolfinx/generation/RectangleMesh.cpp [hunk 1]
```diff
@@ -25,7 +24,7 @@ mesh::Mesh build_tri(MPI_Comm comm,
                      std::array<std::size_t, 2> n,
                      const mesh::GhostMode ghost_mode,
                      const mesh::CellPartitionFunction& partitioner,
-                     const std::string& diagonal)
+                     const generation::DiagonalType diagonal)
```
Comment (garth-wells) [inline cpp/dolfinx/generation/RectangleMesh.cpp:27]: In the DOLFINx style we don't normally use `const` when passing enums.

File: cpp/dolfinx/generation/RectangleMesh.cpp [hunk 2]
```diff
@@ -82,7 +74,7 @@ mesh::Mesh build_tri(MPI_Comm comm,
 
   // Create vertices and cells
   std::size_t nv, nc;
-  if (diagonal == "crossed")
+  if (diagonal == DiagonalType::crossed)
```
Comment (garth-wells) [inline cpp/dolfinx/generation/RectangleMesh.cpp:77]: The `if`s can be changed to switches now that we have enums.

File: cpp/dolfinx/generation/RectangleMesh.h [hunk 3]
```diff
@@ -12,6 +12,21 @@
 #include <mpi.h>
 #include <string>
 
+namespace dolfinx::generation
+{
+
+/// Enum for different diagonal types
+enum class DiagonalType : int
```
Comment (garth-wells) [inline cpp/dolfinx/generation/RectangleMesh.h:19]: Is `int` required here?

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/generation/RectangleMesh.cpp (hunk 1 above)
Comment (jorgensd) [reply cpp/dolfinx/generation/RectangleMesh.cpp:27]: We use it for GhostMode in this function. Do you want me to change that as well?

File: cpp/dolfinx/generation/RectangleMesh.h (hunk 3 above)
Comment (jorgensd) [reply cpp/dolfinx/generation/RectangleMesh.h:19]: I could not get the code to compile without int

**Review by garth-wells — APPROVED** [review]

**Pull request merged**